ES|QL: Remove redundant sorts from execution plan

luigidellaquila opened the pull request:
Fixes: https://github.com/elastic/elasticsearch/issues/120817
Fixes: https://github.com/elastic/elasticsearch/issues/120803

When we have multiple SORT commands in a query, some of them could be redundant.

Eg. 

```
from test | sort x | mv_expand x | sort y | limit 1

from test | sort x | lookup join lookup on x | sort y | limit 1
```

in both cases, the first SORT is practically irrelevant, as the second SORT will reorder the results, and the commands in-between don't rely on the order of results.

In addition, since the second SORT "absorbes" the LIMIT, the first SORT will remain unbounded, and it is problematic, since ES|QL does not support unbounded sort.

This change adds an optimization rule that finds and removes redundant sorts.

A SORT is redundant if all the following conditions are met:
- it is before another SORT.
- all the commands between the two SORTs do not rely on the order of results.

Eg. in `FROM idx | ... | SORT x | <command1> | ... | <commandN> | SORT y`, `SORT x` can be considered redundant:
- because `SORT y` will reorder the results
- but only if none of `<command1>`, `<commandN>` rely on the order of results

Types of commands that could rely on input order are:
- LIMIT: it takes first N results, and if the input is ordered, that order matters
- EVAL with window functions: eg. functions that, for a record, also consider the content of previous/next N records
- STREAMSTATS
- any commands with an inner limit (eg. like MV_EXPAND before [this change](https://github.com/elastic/elasticsearch/pull/120411))

~~At this stage we don't support window functions and STREAMSTATS, but in future we could, so to avoid regressions we consider EVAL and STATS as relevant for order for now.~~

Today we don't support window functions and STREAMSTATS, and we already had rules that do the same thing and don't consider these cases (the old `PruneOrderByBeforeStats` rule), so we'll consider EVAL, FILTER and STATS safe for now.

~~Given this limitation, `AddDefaultTopN` is still in place to make a couple of tests pass, but as reported [here](https://github.com/elastic/elasticsearch/issues/120803), it is most likely incorrect and needs a double-check.~~

Since in some cases we can still have unbounded sorts (eg. `| sort | mv_expand | where`), we also let OrderBy implement `PostOptimizationVerificationAware` and fail validation in a more graceful way.

luigidellaquila committed:
Remove redundant sorts from execution plan

elasticsearchmachine commented:
Hi @luigidellaquila, I've created a changelog YAML for you.

luigidellaquila committed:
Update docs/changelog/121156.yaml

luigidellaquila committed:
More tests

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/esql/remove_redundant_sort' into esql/remove_redundant_sort

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

luigidellaquila committed:
Verify OrderBy after optimization

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -1839,10 +1839,9 @@ public void testCombineOrderByThroughFilter() {
      * Expected
-     * TopN[[Order[first_name{f}#170,ASC,LAST]],1000[INTEGER]]
-     *  \_MvExpand[first_name{f}#170]
-     *    \_TopN[[Order[emp_no{f}#169,ASC,LAST]],1000[INTEGER]]
```
Notice that here you are losing a performance aspect: `mv_expand` is applied on all documents. The idea with a sort (and the **default** pushed down `limit`) in front of `mv_expand` was to have fewer documents to expand from because the expansion is, theoretically, creating even more rows. Before this PR, each node was getting 1000 docs, mv_expanded them and a final sort performed. With this PR, each node is expanding everything, then sorting.

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/RemoveRedundantSort.java:
```diff
@@ -0,0 +1,88 @@
+ *
+ * This rule finds and removes redundant SORTs, making the plan executable.
+ */
+public class RemoveRedundantSort extends OptimizerRules.OptimizerRule<TopN> {
```
It makes sense to combine this with `PruneOrderByBeforeStats` since the logic is similar.
Remove -> Prune

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/RemoveRedundantSort.java:
```diff
@@ -0,0 +1,88 @@
+                    continue;
+                }
+            } else if (p instanceof Join lj) {
+                p = lj.left();
+                // TODO do it also on the right-hand side?
+                continue;
```
Make the function recursive and let it run on both sides.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/RemoveRedundantSort.java:
```diff
@@ -0,0 +1,88 @@
+        });
+    }
+
+    private OrderBy findRedundantSort(TopN plan) {
```
Return a list to perform only one modification (and traversals) on the tree.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/RemoveRedundantSort.java:
```diff
@@ -0,0 +1,88 @@
+public class RemoveRedundantSort extends OptimizerRules.OptimizerRule<TopN> {
+
+    @Override
+    protected LogicalPlan rule(TopN plan) {
+        OrderBy redundant = findRedundantSort(plan);
+        if (redundant == null) {
+            return plan;
+        }
+        return plan.transformDown(p -> {
+            if (p == redundant) {
+                return redundant.child();
+            }
+            return p;
+        });
```
A bottom up traversal would help collect all pruned sorts in one traversal instead of a top-down per sort:
- those that occur before stats
- those that occur before other sorts

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/OrderBy.java:
```diff
@@ -109,4 +110,9 @@ public void postAnalysisVerification(Failures failures) {
+
+    @Override
+    public void postOptimizationVerification(Failures failures) {
+        failures.add(fail(this, "The query cannot be executed because it would require unbounded sort"));
```
I'd rephrase the error message to include an action item for the user: `Unbounded sort not supported yet, please add a limit"

costin reviewed (commented):
It's great to allow more queries - left a few comments regarding the rule implementation.

luigidellaquila committed:
Implement review suggestions

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/esql/remove_redundant_sort' into esql/remove_redundant_sort

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

luigidellaquila committed:
Remove AddDefaultTopN rule

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java:
```diff
@@ -0,0 +1,116 @@
+                    || unary instanceof RegexExtract
+                    || unary instanceof InlineStats
+                    || unary instanceof Lookup
+                // IMPORTANT
+                // If we introduce window functions or order-sensitive aggs (eg. STREAMSTATS),
+                // the previous sort could actually become relevant
+                // so we have to be careful with plans that could use them, ie. the following
+                    || unary instanceof Filter
+                    || unary instanceof Eval
+                    || unary instanceof Aggregate) {
```
This is safe _now_, as long as we don't have window functions (and, in general, functions/aggs that rely on the order of the input).
We can decide to keep it like this for now and review it when we introduce such capabilities, or we can decide to be more paranoid about future regressions and discard such cases, but in this case we won't be able to completely avoid unbounded sorts.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
Let's try to come up with situations where an unbounded sort is unavoidable, and let's add tests for it. I have two reproducers at least:

```
curl -u elastic:password -H "Content-Type: application/json" "127.0.0.1:9200/_query?format=txt" -d '
{
  "query": "row x = [1,2,3], y = 1 | sort y | mv_expand x | where x > 2"
}'
{"error":{"root_cause":[{"type":"verification_exception","reason":"Found 1 problem\nline 1:26: Unbounded sort not supported yet, please add a limit"}],"type":"verification_exception","reason":"Found 1 problem\nline 1:26: Unbounded sort not supported yet, please add a limit"},"status":400}
```

alex-spies replied:
```
curl -u elastic-admin:elastic-password -H "Content-Type: application/json" "127.0.0.1:9200/languages" -XPUT -d '{
  "mappings": {        
        "properties": {"language_code": {"type": "integer"}, "name": {"type": "keyword"}}}, "settings": {"index": {"mode": "lookup"}}}'

curl -u elastic:password -H "Content-Type: application/json" "127.0.0.1:9200/_query?format=txt" -d '
{
  "query": "row language_code = 1, name = \"foo\" | sort name | lookup join languages on language_code | where name == \"English\""
}'

{"error":{"root_cause":[{"type":"verification_exception","reason":"Found 1 problem\nline 1:39: Unbounded sort not supported yet, please add a limit"}],"type":"verification_exception","reason":"Found 1 problem\nline 1:39: Unbounded sort not supported yet, please add a limit"},"status":400}
```

alex-spies replied:
The latter _can_ actually be solved by pushing down the lookup join past the `sort` - which'd require ugly renaming due to shadowing. (One of the lookup fields is called `name`, too.) But pushing down past sorts is not always a good idea. For instance, in something like `... | SORT foo | LOOKUP JOIN idx ON field | LIMIT 10`, we really want the limit to be duplicated downwards, not the sort to be pulled upwards.

Let's open a follow-up issue to investigate these situations.

luigidellaquila replied:
Thanks Alex, I'm adding more tests

I also created an issue for the case above: https://github.com/elastic/elasticsearch/issues/121884

alex-spies reviewed (commented):
I have 2 small testing suggestions and examples for when unbounded sorts still happen.

Important to point out: with this PR, we won't consider remaining unbounded sorts as bugs anymore, and may miss out on their visibility via 500 status codes. (Relevant e.g. for Serverless.)
Because of this, I think we should point this out in our [limitations](https://www.elastic.co/guide/en/elasticsearch/reference/master/esql-limitations.html), mention that users can help themselves via adding a `LIMIT` + are welcome to open a bug report in case they find an unbounded sort that is actually bounded.

alex-spies reviewed (commented):
Reviewed everything but the tests. The main rule implementation looks to be correct to me, but I have minor suggestions on how to maybe make it more robust for the future.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/OrderBy.java:
```diff
@@ -109,4 +110,9 @@ public void postAnalysisVerification(Failures failures) {
+
+    @Override
+    public void postOptimizationVerification(Failures failures) {
+        failures.add(fail(this, "Unbounded sort not supported yet, please add a limit"));
```
nit: maybe we could include the source in the error message to make this easier for users to understand.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java:
```diff
@@ -0,0 +1,116 @@
+                result.put(ob, null);
+                p = ob.child();
+            } else if (p instanceof UnaryPlan unary) {
+                if (unary instanceof Project
```
+1 on a marker interface (say `SortAware`) under capabilities package so that commands can signal this behavior and the rule can pick it up without extra code modifications.
The rule would then perform a top-down transform and keep only the first sort for a `SortAware` (by default Limit to form topN).

P.S. LogicalPlan is a base class, utility methods (assuming they are used beyond just one rule) should be added elsewhere under `PlannerUtils` or `LogicalPlans`.

costin reviewed (commented):
"what Alex said"

luigidellaquila committed:
More tests

luigidellaquila committed:
Better error message

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/OrderBy.java:
```diff
@@ -109,4 +110,9 @@ public void postAnalysisVerification(Failures failures) {
+
+    @Override
+    public void postOptimizationVerification(Failures failures) {
+        failures.add(fail(this, "Unbounded sort not supported yet, please add a limit"));
```
👍 
BWC won't be a problem, right? We are never sending OrderExec to data nodes afaik (we only send a FragmentExec with a logical plan)

luigidellaquila committed:
Delete OrderExec

luigidellaquila committed:
Add SortAware interface

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    private IdentityHashMap<OrderBy, Void> findRedundantSort(LogicalPlan plan) {
+        List<LogicalPlan> toCheck = new ArrayList<>();
```
With the addition of `SortAware` interface this method is becoming much smaller and easier to understand.

If you don't mind, I'd prefer to keep the implicit recursion; the difference in terms of readability is really minimal now and it makes me more comfortable in terms of stack depth.

luigidellaquila commented:
Thanks for the feedback @alex-spies @costin, I added another round of changes that should address your comments

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -7270,4 +7233,301 @@ public void testFunctionNamedParamsAsFunctionArgument() {
+    }
+
+    public void testUnboundedSortSimple() {
+        var query = """
+              ROW x = [1,2,3], y = 1
+              | SORT y
+              | MV_EXPAND x
+              | WHERE x > 2
+            """;
```
suggestion: Could we have a simple test like this but for LOOKUP JOIN? This test is a nice minimal example.

alex-spies reviewed (approved):
This is super nice. I have only minor suggestions. Thanks Luigi!

alex-spies reviewed (commented):
nit: maybe it makes sense to move tests related to the new optimizer rule into a dedicated file. The LogicalPlanOptimizerTests file is already huge and it's hard to see what's tested and not.

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/SortAware.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * This interface is intended to check redundancy of a previous SORT.
+ *
+ * Eg. if a MY_COMMAND that implements this interface is used after a sort, and if dependsOnInputOrder() = false,
+ * then we can assume that
+ * <p>
+ * <code>
+ * | SORT x | MY_COMMAND
+ * </code>
+ * <p>
+ * is equivalent to
+ * <p>
+ * <code>
+ * | MY_COMMAND
+ * </code>
+ * <p>
```
Good point, let me clarify it

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/SortAware.java:
```diff
@@ -0,0 +1,38 @@
+ * In all the other cases, eg. if the command does not implement this interface, or if dependsOnInputOrder() = true
+ * then we assume that the SORT is still relevant and cannot be pruned.
+ */
+public interface SortAware {
```
I agree that the name could be more explicative, but that DoubleSort makes me think about numeric sorting ("double" as a numeric type) and I'm afraid it will create more confusion.

I'm not completely against it though. Up to you folks

alex-spies replied:
That's super bikeshed. Please go ahead with the name you prefer :)

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/SortAware.java:
```diff
@@ -0,0 +1,38 @@
+package org.elasticsearch.xpack.esql.plan.logical;
+
+/**
+ * This interface is intended to check redundancy of a previous SORT.
```
Uhm... this is not completely true.
Eg. in `SORT foo | STATS count(*)` we have only one sort, and the same is with `SORT foo | AnySortAware | STATS count(*)`

The point is more that the command performs some kind of context-free elaboration, but probably it complicates things even more

alex-spies replied:
Hm, good point. This may still be the best way to formalize it, however, because
```STATS ...` is (for now) equivalent to `SORT ... | STATS ...` - and therefore, we can infer that the equivalence `SORT | MY_COMMAND | SORT == MY_COMMAND | SORT` is sufficient to prune a `SORT` that's multiple `SortAware`s upstream from a `STATS`.

This equivalence is also necessary, because for any `SortAware`, we will totally turn `SORT | MY_COMMAND | SORT` into `MY_COMMAND | SORT`.

Sorry for the abstract nonsense, but I'm just more at peace if I know that the contract allows us to prove the correctness of the optimizer rule.

luigidellaquila replied:
Let's give it another try. I pushed some changes, probably too verbose, but more comprehensive

luigidellaquila committed:
Better description for SortAware

luigidellaquila commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -7270,4 +7233,301 @@ public void testFunctionNamedParamsAsFunctionArgument() {
+     *             | \_EsRelation[test][_meta_field{f}#26, emp_no{f}#20, first_name{f}#21, ..]
+     *             \_EsRelation[languages_lookup][LOOKUP][language_code{f}#31]
+     */
+    public void testRedundantSortOnMvExpandJoinEnrichGrokDissect() {
```
Adding one.

> nit: maybe it makes sense to move tests related to the new optimizer rule into a dedicated file.

Let's do it with a separate PR. LogicalPlanOptimizerTests has quite some initialization logic, it will need some refactoring.

luigidellaquila committed:
More tests

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/SortAware.java:
```diff
@@ -0,0 +1,38 @@
+ * </code>
+ * <p>
+ *
+ * In all the other cases, eg. if the command does not implement this interface, or if dependsOnInputOrder() = true
```
Sorting is expensive in a distributed system and commands and users should not depend on it.
At the moment, only Limit and the upcoming change point depend on it, all the other commands do not - hence my preference to opt in. That is the default is optimized for the majority of use-cases.

If it's opt in, the presence of the (marker) interface (SortAware) is enough - no need to implement a method, it's redundant.
If it's opt out, use the same marker interface but change its name to reflect the intent, e.g. `SortIgnorant`, `SortIgnorant` and simply use its presence.

alex-spies replied:
> Sorting is expensive in a distributed system and commands and users should not depend on it.
At the moment, only Limit and the upcoming change point depend on it, all the other commands do not - hence my preference to opt in. That is the default is optimized for the majority of use-cases.

@luigidellaquila and I discussed this and we both think that making the optimization opt-out is a bad idea.

Personally, I have a somewhat strong opinion here. I believe that we'd be really doing ourselves a disservice by making the optimization opt-out as this will, per default, silently break correct queries whenever we add a new command. One such command was just merged (https://github.com/elastic/elasticsearch/pull/120998) today.

The key part here is the "silent". It's silent in the code by being opt-out, but more importantly it's silent to users when they get wrong query results. This increases the risk of rolling out buggy commands whose output doesn't make sense and finding the bug in production at a much later point in time.

Forgetting to ask for the opt-out marker when reviewing a PR with a new command is also something that's really easy to do, resp. the number of things that a reviewer needs to know and look out for is increased. And even for new commands that should be fine with removing upstream sorts, I think it's still better if the first implementation doesn't automatically enable the optimization, so that we can enable it later while simultaneously adding tests.

On the other hand, making the optimization opt-in only increases the verbosity a little bit because most plan nodes actually do opt in. I believe the trade offs here are therefore very much in favor of the opt-in solution. And if we want to keep verbosity down, there's still the possibility to create a new subclass of `UnaryPlan` called `StreamingPlan`, opt this one in, and have all our commands that operate in a row-based manner inherit from that.

> If it's opt in, the presence of the (marker) interface (SortAware) is enough - no need to implement a method, it's redundant.
++, @luigidellaquila and I agreed on going this route.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * breadth-first recursion to find redundant SORTs in the children tree.
+     */
+    private IdentityHashMap<OrderBy, Void> findRedundantSort(LogicalPlan plan) {
```
Hide the implementation by returning the keySet as a `Collection<OrderBy>` on which call `contains`

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java:
```diff
@@ -0,0 +1,84 @@
+                return plan;
+            }
+            return plan.transformDown(p -> {
+                if (redundant.containsKey(p)) {
+                    return ((OrderBy) p).child();
+                }
+                return p;
```
` -> collection.contains(p) ? ((UnaryPlan) p).child() ? p`

costin reviewed (approved):
Small tweaks. LGTM

luigidellaquila committed:
SortAware -> SortAgnostic and simplify

luigidellaquila committed:
Merge branch 'main' into esql/remove_redundant_sort

luigidellaquila commented:
Thanks @costin 

>Sorting is expensive in a distributed system and commands and users should not depend on it.
At the moment, only Limit and the upcoming change point depend on it, all the other commands do not - hence my preference to opt in. That is the default is optimized for the majority of use-cases.

I'd prefer to keep it safe and have the optimization happen only when we know it's safe. ES|QL is evolving fast, and the risk of breaking things is pretty high

>If it's opt in, the presence of the (marker) interface (SortAware) is enough - no need to implement a method, it's redundant.
If it's opt out, use the same marker interface but change its name to reflect the intent, e.g. SortIgnorant, SortIgnorant and simply use its presence.

👍 I renamed it `SortAgnostic` (it's the best I could think of, naming is hard...) and I removed the method (we don't need it for now, we'll add it in the future if we want more control on single commands)

costin commented:
> SortAgnostic
👍

luigidellaquila commented:
Thanks folks!
Merging

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.18  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |
| ❌ |  9.0  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 121156`
